frame/democracy/src/lib.rs

```diff
@@ -0,0 +1,2084 @@
+		pub fn second(
+			origin: OriginFor<T>,
+			#[pallet::compact] proposal: PropIndex,
+			#[pallet::compact] seconds_upper_bound: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let seconds = Self::len_of_deposit_of(proposal).ok_or(Error::<T>::ProposalMissing)?;
+			ensure!(seconds <= seconds_upper_bound, Error::<T>::WrongUpperBound);
+			let mut deposit = Self::deposit_of(proposal).ok_or(Error::<T>::ProposalMissing)?;
+			T::NativeCurrency::hold(&who, deposit.1)?;
+			deposit.0.push(who);
+			<DepositOf<T>>::insert(proposal, deposit);
+			Ok(())
+		}
+
+		/// Vote in a referendum. If `vote.is_aye()`, the vote is to enact the proposal;
+		/// otherwise it is a vote to keep the status quo.
+		///
+		/// The dispatch origin of this call must be _Signed_.
+		///
+		/// - `ref_index`: The index of the referendum to vote for.
+		/// - `vote`: The vote configuration.
+		///
+		/// Weight: `O(R)` where R is the number of referendums the voter has voted on.
+		#[pallet::weight(
+			T::WeightInfo::vote_new(T::MaxVotes::get())
+				.max(T::WeightInfo::vote_existing(T::MaxVotes::get()))
+		)]
+		pub fn vote(
+			origin: OriginFor<T>,
+			#[pallet::compact] ref_index: ReferendumIndex,
+			vote: AccountVote<BalanceOf<T>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			Self::try_vote(&who, ref_index, vote)
+		}
+
+		/// Schedule an emergency cancellation of a referendum. Cannot happen twice to the same
+		/// referendum.
+		///
+		/// The dispatch origin of this call must be `CancellationOrigin`.
+		///
+		/// -`ref_index`: The index of the referendum to cancel.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::weight((T::WeightInfo::emergency_cancel(), DispatchClass::Operational))]
+		pub fn emergency_cancel(
+			origin: OriginFor<T>,
+			ref_index: ReferendumIndex,
+		) -> DispatchResult {
+			T::CancellationOrigin::ensure_origin(origin)?;
+
+			let status = Self::referendum_status(ref_index)?;
+			let id = status.proposal_id;
+			ensure!(!<Cancellations<T>>::contains_key(&id), Error::<T>::AlreadyCanceled);
+
+			<Cancellations<T>>::insert(id, true);
+			Self::internal_cancel_referendum(ref_index);
+			Ok(())
+		}
+
+		/// Schedule a referendum to be tabled once it is legal to schedule an external
+		/// referendum.
+		///
+		/// The dispatch origin of this call must be `ExternalOrigin`.
+		///
+		/// - `proposal_hash`: The preimage hash of the proposal.
+		///
+		/// Weight: `O(V)` with V number of vetoers in the blacklist of proposal.
+		///   Decoding vec of length V. Charged as maximum
+		#[pallet::weight(T::WeightInfo::external_propose(MAX_VETOERS))]
+		pub fn external_propose(
+			origin: OriginFor<T>,
+			proposal_hash: T::Hash,
+			asset_id: T::AssetId,
+		) -> DispatchResult {
+			T::ExternalOrigin::ensure_origin(origin)?;
+			ensure!(!<NextExternal<T>>::exists(), Error::<T>::DuplicateProposal);
+			let id = ProposalId { hash: proposal_hash, asset_id };
```

Comment:
is assets exists?

Reply:
It's upto to the user to make sure that the combination makes sense. We're not checking if asset id exists.

Reply:
so he can create vote, than somebody creates asset. and vote will happen on asset which not supposed to be happen?

Reply:
Only during proposal asset_id can be given, while voting they can only vote with the asset id meant for that proposal( or referendum).

Reply:
you give asset_id during proposal. id is not yet exists. somebody observers that offchain. and adds asset_id on their own under their control. and votes against or for  or what ever.